### PR TITLE
Implement support for Fullscreen API

### PR DESCRIPTION
**Pull request by Emdek**

### CODE REVIEW

File: Source/WebKit/qt/Api/qwebelement.cpp [hunk 1]
```diff
@@ -1502,6 +1502,22 @@ void QWebElement::render(QPainter* painter, const QRect& clip)
     context.restore();
 }
 
+void QWebElement::enterFullScreen()
+{
+#if ENABLE(FULLSCREEN_API)
+    m_element->document().webkitWillEnterFullScreenForElement(m_element);
```
Comment (annulen) [inline Source/WebKit/qt/Api/qwebelement.cpp]: I think WillEnter should be called before resize and DidEnter after

File: Source/WebKit/qt/Api/qwebelement.cpp [hunk 2]
```diff
@@ -1502,6 +1502,22 @@ void QWebElement::render(QPainter* painter, const QRect& clip)
...
+    m_element->document().webkitDidEnterFullScreenForElement(m_element);
+#endif
+}
+
+void QWebElement::exitFullScreen()
+{
+#if ENABLE(FULLSCREEN_API)
+    m_element->document().webkitWillExitFullScreenForElement(m_element);
```
Comment (annulen) [inline Source/WebKit/qt/Api/qwebelement.cpp]: Ditto

File: Source/WebKit/qt/Api/qwebelement.h [hunk 3]
```diff
@@ -147,11 +149,17 @@ class QWEBKIT_EXPORT QWebElement {
...
+    void exitFullScreen();
+
     static QWebElement enclosingElement(WebCore::Node*);
 
+    friend class WebCore::ChromeClientQt;
     friend class DumpRenderTreeSupportQt;
     friend class QWebFrameAdapter;
+    friend class QWebPageAdapter;
     friend class QWebElementCollection;
+    friend class QWebFullScreenRequest;
```
Comment (annulen) [inline Source/WebKit/qt/Api/qwebelement.h]: Sort order is not alphabetic

File: Source/WebKit/qt/Api/qwebsettings.cpp [hunk 4]
```diff
@@ -565,6 +568,7 @@ QWebSettings::QWebSettings()
     d->attributes.insert(QWebSettings::ScrollAnimatorEnabled, false);
     d->attributes.insert(QWebSettings::CaretBrowsingEnabled, false);
     d->attributes.insert(QWebSettings::NotificationsEnabled, true);
+    d->attributes.insert(QWebSettings::FullScreenSupportEnabled, false);
```
Comment (annulen) [inline Source/WebKit/qt/Api/qwebsettings.cpp]: Does "false" match old behavior? IIRC fullscreen for `<video>` was available by default

File: Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp [hunk 5]
```diff
@@ -708,6 +708,23 @@ void ChromeClientQt::exitVideoFullscreenForVideoElement(HTMLVideoElement& videoE
...
 
+#if ENABLE(FULLSCREEN_API)
+bool ChromeClientQt::supportsFullScreenForElement(const Element*, bool withKeyboard)
+{
+    return !withKeyboard;
+}
+
+void ChromeClientQt::enterFullScreenForElement(Element* element)
+{
+    m_webPage->fullScreenRequested(QUrl(element->document().url()), QWebElement(element), true);
```
Comment (annulen) [inline Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp]: Should work just as `m_webPage->fullScreenRequested(element->document().url(), QWebElement(element), true);

File: Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp [hunk 6]
```diff
@@ -708,6 +708,23 @@ void ChromeClientQt::exitVideoFullscreenForVideoElement(HTMLVideoElement& videoE
...
+}
+
+void ChromeClientQt::exitFullScreenForElement(Element* element)
+{
+    m_webPage->fullScreenRequested(QUrl(element->document().url()), QWebElement(element), false);
```
Comment (annulen) [inline Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp]: Ditto

File: Source/WebKit/qt/WidgetApi/qwebfullscreenrequest.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+#include "config.h"
+#include "qwebfullscreenrequest.h"
+
+#include "qwebpage_p.h"
+
+#include <QtCore/qobject.h>
+
+QT_BEGIN_NAMESPACE
+
+QWebFullScreenRequest::QWebFullScreenRequest(QWebPage *page, const QUrl &origin, const QWebElement &element, bool fullscreen)
```
Comment (annulen) [inline Source/WebKit/qt/WidgetApi/qwebfullscreenrequest.cpp]: For some reason `check-webkit-style` seems to skip this line, incorrect positions of * and &

File: Source/WebKit/qt/WidgetApi/qwebfullscreenrequest.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#include <QtCore/qpointer.h>
+#include <QtCore/qurl.h>
+#include <QtWebKit/qwebelement.h>
+#include <QtWebKit/qwebkitglobal.h>
+
+QT_BEGIN_NAMESPACE
+class QWebPage;
+
+class QWEBKITWIDGETS_EXPORT QWebFullScreenRequest {
+    Q_GADGET
```
Comment (annulen) [inline Source/WebKit/qt/WidgetApi/qwebfullscreenrequest.h]: Let it be Q_OBJECT

File: Source/WebKit/qt/WidgetApi/qwebpage_p.cpp [hunk 9]
```diff
@@ -53,6 +54,12 @@ QWebPagePrivate::~QWebPagePrivate()
 
 QWebFramePrivate::~QWebFramePrivate() = default;
 
+void QWebPagePrivate::fullScreenRequested(const QUrl &origin, const QWebElement &element, bool fullscreen)
```
Comment (annulen) [inline Source/WebKit/qt/WidgetApi/qwebpage_p.cpp]: Style again

File: Source/WebKit/qt/Api/qwebsettings.cpp (hunk 4 above)
Comment (Emdek) [inline Source/WebKit/qt/Api/qwebsettings.cpp]: It's a default value in Settings.in and I don't see any other call to change it in Qt port other than our new attribute.
They seem to be completely separate.

**Comment by annulen** [comment]
With patch
```diff
diff --git a/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp b/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp
index a32b87d..6fc0768 100644
--- a/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp
+++ b/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp
@@ -716,12 +716,22 @@ bool ChromeClientQt::supportsFullScreenForElement(const Element*, bool withKeybo
 
 void ChromeClientQt::enterFullScreenForElement(Element* element)
 {
+    m_fullScreenElement = element;
+
+    qDebug() << Q_FUNC_INFO << 1 << element;
     m_webPage->fullScreenRequested(element->document().url(), QWebElement(element), true);
+    qDebug() << Q_FUNC_INFO << m_fullScreenElement.get();
 }
 
-void ChromeClientQt::exitFullScreenForElement(Element* element)
+void ChromeClientQt::exitFullScreenForElement(Element*)
 {
-    m_webPage->fullScreenRequested(element->document().url(), QWebElement(element), false);
+    qDebug() << Q_FUNC_INFO << m_fullScreenElement.get();
+    // The element passed into this function is not reliable, i.e. it could
+    // be null. In addition the parameter may be disappearing in the future.
+    // So we use the reference to the element we saved above.
+    ASSERT(m_fullScreenElement);
+    m_webPage->fullScreenRequested(m_fullScreenElement->document().url(), QWebElement(m_fullScreenElement.get()), false);
+    m_fullScreenElement = nullptr;
 }
 #endif
 
diff --git a/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.h b/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.h
index 2e93e07..69edbcf 100644
--- a/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.h
+++ b/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.h
@@ -226,6 +226,8 @@ public:
     mutable QtPlatformPlugin m_platformPlugin;
 
     std::unique_ptr<TextureMapperLayerClientQt> m_textureMapperLayerClient;
+
+    RefPtr<WebCore::Element> m_fullScreenElement;
 };
 }
 
diff --git a/Tools/DumpRenderTree/qt/DumpRenderTreeQt.cpp b/Tools/DumpRenderTree/qt/DumpRenderTreeQt.cpp
index b730df6..b739805 100755
--- a/Tools/DumpRenderTree/qt/DumpRenderTreeQt.cpp
+++ b/Tools/DumpRenderTree/qt/DumpRenderTreeQt.cpp
@@ -155,6 +155,8 @@ WebPage::WebPage(QObject* parent, DumpRenderTree* drt)
 
     connect(this, SIGNAL(featurePermissionRequested(QWebFrame*, QWebPage::Feature)), this, SLOT(requestPermission(QWebFrame*, QWebPage::Feature)));
     connect(this, SIGNAL(featurePermissionRequestCanceled(QWebFrame*, QWebPage::Feature)), this, SLOT(cancelPermission(QWebFrame*, QWebPage::Feature)));
+
+    connect(this, &QWebPage::fullScreenRequested, this, &WebPage::requestFullScreen);
 }
 
 WebPage::~WebPage()
@@ -192,6 +194,8 @@ void WebPage::resetSettings()
     settings()->resetAttribute(QWebSettings::CSSGridLayoutEnabled);
     settings()->resetAttribute(QWebSettings::AcceleratedCompositingEnabled);
 
+    settings()->setAttribute(QWebSettings::FullScreenSupportEnabled, true);
+
     m_drt->testRunner()->setCaretBrowsingEnabled(false);
     m_drt->testRunner()->setAuthorAndUserStylesEnabled(true);
     m_drt->jscTestRunner()->setDefersLoading(false);
@@ -273,6 +277,12 @@ void WebPage::permissionSet(QWebPage::Feature feature)
     }
 }
 
+void WebPage::requestFullScreen(QWebFullScreenRequest request)
+{
+    qDebug() << Q_FUNC_INFO << request.toggleOn();
+    request.accept();
+}
+
 // FIXME (119591): Make this match other platforms better.
 static QString urlSuitableForTestResult(const QString& url)
 {
diff --git a/Tools/DumpRenderTree/qt/DumpRenderTreeQt.h b/Tools/DumpRenderTree/qt/DumpRenderTreeQt.h
index be1559d..72344ba 100644
--- a/Tools/DumpRenderTree/qt/DumpRenderTreeQt.h
+++ b/Tools/DumpRenderTree/qt/DumpRenderTreeQt.h
@@ -213,6 +213,7 @@ public:
 public Q_SLOTS:
     void requestPermission(QWebFrame* frame, QWebPage::Feature feature);
     void cancelPermission(QWebFrame* frame, QWebPage::Feature feature);
+    void requestFullScreen(QWebFullScreenRequest);
 
 protected:
     bool acceptNavigationRequest(QWebFrame* frame, const QNetworkRequest& request, NavigationType type);
```
still get

```
Regressions: Unexpected text-only failures (3)                                                                                                    
  fullscreen/full-screen-placeholder.html [ Failure ]                                                                                             
  fullscreen/full-screen-with-flex-item.html [ Failure ]                                                                                          
  fullscreen/video-controls-override.html [ Failure ]                                                                                             
                                                                                                                                                  
Regressions: Unexpected image-only failures (1)                                                                                                   
  fullscreen/parent-flow-inline-with-block-child.html [ ImageOnlyFailure ]                                                                        
                                                                                                                                                  
Regressions: Unexpected crashes (1)                                                                                                               
  fullscreen/full-screen-restrictions.html [ Crash ]                                                                                              
                                                                                                                                                  
Regressions: Unexpected timeouts (5)                                                                                                              
  fullscreen/full-screen-exit-when-popup.html [ Timeout ]                                                                                         
  fullscreen/full-screen-inline-split-crash.html [ Timeout ]                                                                                      
  fullscreen/full-screen-keyboard-enabled.html [ Timeout ]                                                                                        
  fullscreen/full-screen-plugin.html [ Timeout ]                                                                                                  
  fullscreen/full-screen-table-section.html [ Timeout ]                                                                                           
                                                                                                                                                  
Regressions: Unexpected missing results (5)                                                                                                       
  fullscreen/full-screen-iframe-zIndex.html [ Missing ]                                                                                           
  fullscreen/full-screen-remove-ancestor-after.html [ Missing ]                                                                                   
  fullscreen/full-screen-stacking-context.html [ Missing ]                                                                                        
  fullscreen/full-screen-zIndex-after.html [ Missing ]                                                                                            
  fullscreen/full-screen-zIndex.html [ Missing ]
```

**Comment by annulen** [comment]
For the record, WK2 has following code in `FrameLoaderClient::dispatchDidStartProvisionalLoad` override
```c++
#if ENABLE(FULLSCREEN_API)
    Element* documentElement = m_frame->coreFrame()->document()->documentElement();
    if (documentElement && documentElement->containsFullScreenElement())
        webPage->fullScreenManager()->exitFullScreenForElement(webPage->fullScreenManager()->element());
#endif
```

**Comment by annulen** [comment]
Superceded by #377

**Pull request closed without merging**